File: frontend/scripts/script-env.js
#!/usr/bin/env node
/**
 * 腳本環境配置模組
 * 提供統一的 Docker 環境檢測和配置管理
 */

const fs = require('fs');
const path = require('path');

class ScriptEnvironment {
  constructor() {
    this.projectRoot = this.findProjectRoot();
    this.dockerConfig = this.detectDockerEnvironment();
    this.envConfig = this.loadEnvironmentConfig();
  }

  /**
   * 尋找項目根目錄
   */
  findProjectRoot() {
    let currentDir = process.cwd();

    while (currentDir !== path.parse(currentDir).root) {
      if (
        fs.existsSync(path.join(currentDir, 'package.json')) ||
        fs.existsSync(path.join(currentDir, 'docker-compose.yml'))
      ) {
        return currentDir;
      }
      currentDir = path.dirname(currentDir);
    }

    return process.cwd();
  }

  /**
   * 檢測 Docker 環境
   */
  detectDockerEnvironment() {
    const checks = {
      // 檔案存在檢查
      dockerFile: fs.existsSync('/.dockerenv'),

      // 環境變數檢查
      nodeEnv: process.env.NODE_ENV === 'docker',
      dockerEnv: process.env.DOCKER_ENV === 'true',

      // 容器名稱檢查
      frontendContainer: process.env.FRONTEND_URL?.includes('://frontend:'),
      backendContainer: process.env.BACKEND_URL?.includes('://backend:'),

      // 主機名檢查
      hostname: process.env.HOSTNAME?.startsWith('stock-insight-'),

      // Docker Compose 服務檢查
      dockerCompose: fs.existsSync(path.join(this.projectRoot, 'docker-compose.yml')),
    };

    const isDocker = Object.values(checks).some((check) => check === true);

    return {
      isDocker,
      checks,
      confidence: Object.values(checks).filter(Boolean).length / Object.keys(checks).length,
    };
  }

  /**
   * 載入環境配置
   */
  loadEnvironmentConfig() {
    const { isDocker } = this.dockerConfig;

    // 基礎配置
    const baseConfig = {
      // 前端配置
      frontend: {
        host: isDocker ? 'frontend' : 'localhost',
        port: parseInt(process.env.FRONTEND_PORT || '5173'),
        protocol: 'http',
      },

      // 後端配置
      backend: {
        host: isDocker ? 'backend' : 'localhost',
        port: parseInt(process.env.BACKEND_PORT || '5001'),
        protocol: 'http',
      },

      // 資料庫配置
      database: {
        host: isDocker ? 'db' : 'localhost',
        port: parseInt(process.env.DB_PORT || '5433'),
        name: process.env.DB_NAME || 'stock_insight',
      },

      // Redis 配置
      redis: {
        host: isDocker ? 'redis' : 'localhost',
        port: parseInt(process.env.REDIS_PORT || '6379'),
      },

      // 通用配置
      timeout: parseInt(process.env.REQUEST_TIMEOUT || '10000'),
      retries: parseInt(process.env.MAX_RETRIES || '3'),
      debug: process.env.DEBUG === 'true',
    };

    // URL 構建
    const urls = {
      frontend:
        process.env.FRONTEND_URL ||
        `${baseConfig.frontend.protocol}://${baseConfig.frontend.host}:${baseConfig.frontend.port}`,

      backend:
        process.env.BACKEND_URL ||
        (isDocker
          ? ''
          : `${baseConfig.backend.protocol}://${baseConfig.backend.host}:${baseConfig.backend.port}`),

      api: isDocker
        ? '/api'
        : `${baseConfig.backend.protocol}://${baseConfig.backend.host}:${baseConfig.backend.port}/api`,
    };

    return {
      ...baseConfig,
      urls,
      isDocker,
      environment: process.env.NODE_ENV || (isDocker ? 'docker' : 'development'),
    };
  }

  /**
   * 獲取完整環境信息
   */
  getEnvironmentInfo() {
    return {
      projectRoot: this.projectRoot,
      docker: this.dockerConfig,
      config: this.envConfig,
    };
  }

  /**
   * 打印環境信息
   */
  printEnvironmentInfo() {
    const { dockerConfig, envConfig } = this;

    console.log('🔍 腳本環境檢測結果:');
    console.log('='.repeat(40));
    console.log(`📍 項目根目錄: ${this.projectRoot}`);
    console.log(
      `🐳 Docker 環境: ${dockerConfig.isDocker ? '是' : '否'} (信心度: ${(dockerConfig.confidence * 100).toFixed(1)}%)`,
    );
    console.log(`🌍 執行環境: ${envConfig.environment}`);
    console.log(`⚙️  設定:`);
    console.log(`   - 前端: ${envConfig.urls.frontend}`);
    console.log(`   - 後端: ${envConfig.urls.backend || '使用代理'}`);
    console.log(`   - API: ${envConfig.urls.api}`);
    console.log(`   - 超時: ${envConfig.timeout}ms`);

    if (envConfig.debug) {
      console.log('\n🔍 詳細檢測結果:');
      Object.entries(dockerConfig.checks).forEach(([key, value]) => {
        console.log(`   - ${key}: ${value ? '✅' : '❌'}`);
      });
    }

    console.log('='.repeat(40));
  }

  /**
   * 靜態方法：快速獲取環境配置
   */
  static getConfig() {
    const env = new ScriptEnvironment();
    return env.getEnvironmentInfo();
  }

  /**
   * 靜態方法：快速檢查是否為 Docker 環境
   */
  static isDocker() {
    const env = new ScriptEnvironment();
    return env.dockerConfig.isDocker;
  }
}

module.exports = {
  ScriptEnvironment,
  getConfig: ScriptEnvironment.getConfig,
  isDocker: ScriptEnvironment.isDocker,
};

// 當作為主模組執行時，顯示環境信息
if (require.main === module) {
  console.log('🧪 測試前端腳本環境配置模組...\n');

  try {
    const env = new ScriptEnvironment();
    env.printEnvironmentInfo();

    console.log('\n✅ 前端環境配置模組測試成功！');
  } catch (error) {
    console.error('❌ 前端環境配置模組測試失敗:', error.message);
    process.exit(1);
  }
}

File: backend/scripts/script_env.py
#!/usr/bin/env python3
"""
後端腳本環境配置模組
提供統一的 Docker 環境檢測和配置管理
"""

import os
import sys
from pathlib import Path
from typing import Any, Dict, Tuple


class ScriptEnvironment:
    """腳本環境配置管理器"""

    def __init__(self):
        self.project_root = self._find_project_root()
        self.docker_config = self._detect_docker_environment()
        self.env_config = self._load_environment_config()

    def _find_project_root(self) -> Path:
        """尋找項目根目錄"""
        current_dir = Path.cwd()

        while current_dir != current_dir.parent:
            if (
                (current_dir / "requirements.txt").exists()
                or (current_dir / "docker-compose.yml").exists()
                or (current_dir / "manage.py").exists()
            ):
                return current_dir
            current_dir = current_dir.parent

        return Path.cwd()

    def _detect_docker_environment(self) -> Dict[str, Any]:
        """檢測 Docker 環境"""
        checks = {
            # 檔案存在檢查
            "docker_file": Path("/.dockerenv").exists(),
            # 環境變數檢查
            "node_env": os.environ.get("NODE_ENV") == "docker",
            "docker_env": os.environ.get("DOCKER_ENV") == "true",
            # 容器名稱檢查
            "frontend_container": "://frontend:" in os.environ.get("FRONTEND_URL", ""),
            "backend_container": "://backend:" in os.environ.get("BACKEND_URL", ""),
            # 主機名檢查
            "hostname": os.environ.get("HOSTNAME", "").startswith("stock-insight-"),
            # Docker Compose 服務檢查
            "docker_compose": (self.project_root / "docker-compose.yml").exists(),
        }

        is_docker = any(checks.values())
        confidence = sum(checks.values()) / len(checks)

        return {"is_docker": is_docker, "checks": checks, "confidence": confidence}

    def _load_environment_config(self) -> Dict[str, Any]:
        """載入環境配置"""
        is_docker = self.docker_config["is_docker"]

        # 基礎配置
        base_config = {
            # 前端配置
            "frontend": {
                "host": "frontend" if is_docker else "localhost",
                "port": int(os.environ.get("FRONTEND_PORT", "5173")),
                "protocol": "http",
            },
            # 後端配置
            "backend": {
                "host": "0.0.0.0" if is_docker else "localhost",
                "port": int(os.environ.get("BACKEND_PORT", "5000")),
                "protocol": "http",
            },
            # 資料庫配置
            "database": {
                "host": "db" if is_docker else "localhost",
                "port": int(os.environ.get("DB_PORT", "1433")),  # MSSQL 預設端口
                "name": os.environ.get("MSSQL_DATABASE", "StockInsight"),
                "user": os.environ.get("MSSQL_USER", "sa"),
                "password": os.environ.get("MSSQL_SA_PASSWORD", "YourStrong!Password123"),
                "driver": "ODBC+Driver+18+for+SQL+Server",
            },
            # Redis 配置
            "redis": {
                "host": "redis" if is_docker else "localhost",
                "port": int(os.environ.get("REDIS_PORT", "6379")),
            },
            # 通用配置
            "timeout": int(os.environ.get("REQUEST_TIMEOUT", "10")),
            "retries": int(os.environ.get("MAX_RETRIES", "3")),
            "debug": os.environ.get("DEBUG", "False").lower() == "true",
        }

        # URL 構建
        urls = {
            "frontend": os.environ.get("FRONTEND_URL")
            or f"{base_config['frontend']['protocol']}://{base_config['frontend']['host']}:{base_config['frontend']['port']}",
            "backend": os.environ.get("BACKEND_URL")
            or f"{base_config['backend']['protocol']}://{base_config['backend']['host']}:{base_config['backend']['port']}",
            "database": os.environ.get("DATABASE_URL")
            or (
                f"mssql+pyodbc://{base_config['database']['user']}:{base_config['database']['password']}@{base_config['database']['host']}:{base_config['database']['port']}/{base_config['database']['name']}?driver={base_config['database']['driver']}&TrustServerCertificate=yes"
                if is_docker
                else "sqlite:///app.db"
            ),
            "redis": os.environ.get("REDIS_URL")
            or f"redis://{base_config['redis']['host']}:{base_config['redis']['port']}/0",
        }

        return {
            **base_config,
            "urls": urls,
            "is_docker": is_docker,
            "environment": os.environ.get("NODE_ENV", "docker" if is_docker else "development"),
        }

    def get_environment_info(self) -> Dict[str, Any]:
        """獲取完整環境信息"""
        return {
            "project_root": str(self.project_root),
            "docker": self.docker_config,
            "config": self.env_config,
        }

    def print_environment_info(self):
        """打印環境信息"""
        docker = self.docker_config
        config = self.env_config

        print("🔍 腳本環境檢測結果:")
        print("=" * 40)
        print(f"📍 項目根目錄: {self.project_root}")
        print(
            f'🐳 Docker 環境: {"是" if docker["is_docker"] else "否"} (信心度: {docker["confidence"] * 100:.1f}%)'
        )
        print(f'🌍 執行環境: {config["environment"]}')
        print(f"⚙️  設定:")
        print(f'   - 前端: {config["urls"]["frontend"]}')
        print(f'   - 後端: {config["urls"]["backend"]}')
        print(f'   - 資料庫: {config["urls"]["database"]}')
        print(f'   - Redis: {config["urls"]["redis"]}')
        print(f'   - 超時: {config["timeout"]}s')

        if config["debug"]:
            print("\n🔍 詳細檢測結果:")
            for key, value in docker["checks"].items():
                print(f'   - {key}: {"✅" if value else "❌"}')

        print("=" * 40)

    @staticmethod
    def get_config() -> Dict[str, Any]:
        """靜態方法：快速獲取環境配置"""
        env = ScriptEnvironment()
        return env.get_environment_info()

    @staticmethod
    def is_docker() -> bool:
        """靜態方法：快速檢查是否為 Docker 環境"""
        env = ScriptEnvironment()
        return env.docker_config["is_docker"]

    @staticmethod
    def get_app_config() -> Tuple[str, int, bool]:
        """獲取應用配置 (host, port, debug) - 向後兼容"""
        env = ScriptEnvironment()
        config = env.env_config
        return (config["backend"]["host"], config["backend"]["port"], config["debug"])


# 向後兼容的便利函數
def is_docker_environment():
    """檢測是否在 Docker 環境中運行 - 向後兼容"""
    return ScriptEnvironment.is_docker()


def get_app_config():
    """根據環境獲取應用配置 - 向後兼容"""
    return ScriptEnvironment.get_app_config()


if __name__ == "__main__":
    # 測試腳本
    env = ScriptEnvironment()
    env.print_environment_info()
